File: 2024/4/part1.py
import re

raw = open('input.dat').readlines()
raw = [line.strip() for line in raw]

n, m = len(raw), len(raw[0])

def check_horizontal(i, j):
    count = 0
    # Check rightward (XMAS →)
    if j + 3 < m:
        if raw[i][j] == 'X' and raw[i][j + 1] == 'M' and raw[i][j + 2] == 'A' and raw[i][j + 3] == 'S':
            count += 1
    # Check leftward (SAMS ←)
    if j - 3 >= 0:
        if raw[i][j] == 'X' and raw[i][j - 1] == 'M' and raw[i][j - 2] == 'A' and raw[i][j - 3] == 'S':
            count += 1
    return count

def check_vertical(i, j):
    count = 0
    # Check downward (XMAS ↓)
    if i + 3 < n:
        if raw[i][j] == 'X' and raw[i + 1][j] == 'M' and raw[i + 2][j] == 'A' and raw[i + 3][j] == 'S':
            count += 1
    # Check upward (SAMS ↑)
    if i - 3 >= 0:
        if raw[i][j] == 'X' and raw[i - 1][j] == 'M' and raw[i - 2][j] == 'A' and raw[i - 3][j] == 'S':
            count += 1
    return count


def check_diagonal(i, j):
    count = 0
    # Check diagonal downward-right (XMAS ↘)
    if i + 3 < n and j + 3 < m:
        if raw[i][j] == 'X' and raw[i + 1][j + 1] == 'M' and raw[i + 2][j + 2] == 'A' and raw[i + 3][j + 3] == 'S':
            count += 1
    # Check diagonal upward-left (SAMS ↖)
    if i - 3 >= 0 and j - 3 >= 0:
        if raw[i][j] == 'X' and raw[i - 1][j - 1] == 'M' and raw[i - 2][j - 2] == 'A' and raw[i - 3][j - 3] == 'S':
            count += 1
    # Check diagonal downward-left (XMAS ↙)
    if i + 3 < n and j - 3 >= 0:
        if raw[i][j] == 'X' and raw[i + 1][j - 1] == 'M' and raw[i + 2][j - 2] == 'A' and raw[i + 3][j - 3] == 'S':
            count += 1
    # Check diagonal upward-right (SAMS ↗)
    if i - 3 >= 0 and j + 3 < m:
        if raw[i][j] == 'X' and raw[i - 1][j + 1] == 'M' and raw[i - 2][j + 2] == 'A' and raw[i - 3][j + 3] == 'S':
            count += 1
    return count


def check_all(i, j):
    return check_horizontal(i, j) + check_vertical(i, j) + check_diagonal(i, j)

cnt = 0
for i, row in enumerate(raw):
    for j, col in enumerate(row):
        cnt += check_all(i, j)

print(cnt)

File: 2024/go/4/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	file, err := os.Open("input.dat")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	matrix := [][]rune{}
	for scanner.Scan() {
		parts := scanner.Text()
		runeSlice := []rune(parts)
		matrix = append(matrix, runeSlice)
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	var cnt int
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[i]); j++ {
			if matrix[i][j] == 'X' {
				cnt += checkAll(matrix, i, j)
			}
		}
	}

	fmt.Println(cnt)
}

func checkHorizontal(matrix [][]rune, i int, j int) int {
	var cnt int
	// Check Left
	if (j-3 >= 0) && (matrix[i][j-1] == 'M') && (matrix[i][j-2] == 'A') && (matrix[i][j-3] == 'S') {
		cnt += 1
	}
	// Check Right
	if (j+3 < len(matrix[i])) && (matrix[i][j+1] == 'M') && (matrix[i][j+2] == 'A') && (matrix[i][j+3] == 'S') {
		cnt += 1
	}
	return cnt
}

func checkVertical(matrix [][]rune, i int, j int) int {
	var cnt int
	// Check Up
	if (i-3 >= 0) && (matrix[i-1][j] == 'M') && (matrix[i-2][j] == 'A') && (matrix[i-3][j] == 'S') {
		cnt += 1
	}
	// Check Down
	if (i+3 < len(matrix)) && (matrix[i+1][j] == 'M') && (matrix[i+2][j] == 'A') && (matrix[i+3][j] == 'S') {
		cnt += 1
	}
	return cnt
}

func checkDiagonal(matrix [][]rune, i int, j int) int {
	var cnt int
	// Check Up Left
	if (i-3 >= 0) && (j-3 >= 0) && (matrix[i-1][j-1] == 'M') && (matrix[i-2][j-2] == 'A') && (matrix[i-3][j-3] == 'S') {
		cnt += 1
	}
	// Check Up Right
	if (i-3 >= 0) && (j+3 < len(matrix[i])) && (matrix[i-1][j+1] == 'M') && (matrix[i-2][j+2] == 'A') && (matrix[i-3][j+3] == 'S') {
		cnt += 1
	}
	// Check Down Left
	if (i+3 < len(matrix)) && (j-3 >= 0) && (matrix[i+1][j-1] == 'M') && (matrix[i+2][j-2] == 'A') && (matrix[i+3][j-3] == 'S') {
		cnt += 1
	}
	// Check Down Right
	if (i+3 < len(matrix)) && (j+3 < len(matrix[i])) && (matrix[i+1][j+1] == 'M') && (matrix[i+2][j+2] == 'A') && (matrix[i+3][j+3] == 'S') {
		cnt += 1
	}
	return cnt
}

func checkAll(matrix [][]rune, i int, j int) int {
	var cnt int
	cnt += checkHorizontal(matrix, i, j)
	cnt += checkVertical(matrix, i, j)
	cnt += checkDiagonal(matrix, i, j)
	return cnt
}
